# 多項ロジットモデル

**多項ロジットモデル（multinomial logt model）**
多項ロジットモデルは、被説明変数が取りうる値が3つ以上で、それらに順序がない場合に適用します。

例えば、移動手段（バス・自動車・電車・徒歩・自転車・その他）を選択肢から選ぶ場合や、消費者が購入する商品を、ブランドA、ブランドB、ブランドCから選択する場合などに適用することができます。


`statsmodels`を用いて推定します。
 


ここではスマートフォンを用いて収集されたセンサーデータから、スマートフォンユーザーの移動手段を推定する例を考えます。

このデータは、13人のボランティア（男性10名、女性3名）のスマートフォンから収集したデータで、徒歩、車、静止状態、電車、バスの5クラスで構成されています。

データ出典: http://cs.unibo.it/projects/us-tm2017/index.html



In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

まずはデータをダウンロードします。

In [2]:
url = "http://cs.unibo.it/projects/us-tm2017/static/dataset/extension/5second/csv/dataset_5secondWindow.tar.gz"
df = pd.read_csv(url, compression='gzip', header=0, sep=',', quotechar='"', on_bad_lines='warn')
#最初の数行を表示させます
pd.set_option('display.max_columns', 100) # すべての列がNotebook上で表示されるように、オプション設定を変更します。
df.head(2)

dataset_5secondWindow.csv       id   time  activityrecognition#0  \
0                        0.0  16170.0   78.0                    NaN   
1                        1.0  15871.0  145.0                    NaN   

   activityrecognition#1  android.sensor.accelerometer#mean  \
0                  100.0                           9.811476   
1                  100.0                           9.939207   

   android.sensor.accelerometer#min  android.sensor.accelerometer#max  \
0                          9.758895                          9.849411   
1                          7.707437                         17.146631   

   android.sensor.accelerometer#std  android.sensor.game_rotation_vector#mean  \
0                          0.014626                                  0.029340   
1                          1.775944                                  0.999925   

   android.sensor.game_rotation_vector#min  \
0                                 0.029014   
1                                 0.999903   

   android.sensor.game_rotation_vector#max  \
0                                 0.029526   
1                                 0.999946   

   android.sensor.game_rotation_vector#std  android.sensor.gravity#mean  \
0                                 0.000119                     9.806650   
1                                 0.000030                     9.806624   

   android.sensor.gravity#min  android.sensor.gravity#max  \
0                    9.806649                    9.806651   
1                    9.806624                    9.806624   

   android.sensor.gravity#std  android.sensor.gyroscope#mean  \
0                4.780692e-07                       0.001651   
1                6.474977e-07                       0.036326   

   android.sensor.gyroscope#min  android.sensor.gyroscope#max  \
0                      0.000000                      0.003533   
1                      0.011669                      0.059388   

   android.sensor.gyroscope#std  android.sensor.gyroscope_uncalibrated#mean  \
0                      0.000737                                    0.016221   
1                      0.020290                                    0.039023   

   android.sensor.gyroscope_uncalibrated#min  \
0                                   0.014172   
1                                   0.014132   

   android.sensor.gyroscope_uncalibrated#max  \
0                                   0.018695   
1                                   0.085494   

   android.sensor.gyroscope_uncalibrated#std  android.sensor.light#mean  \
0                                   0.000982                        0.0   
1                                   0.018629                        0.0   

   android.sensor.light#min  android.sensor.light#max  \
0                       0.0                       0.0   
1                       0.0                       0.0   

   android.sensor.light#std  android.sensor.linear_acceleration#mean  \
0                       0.0                                 0.020978   
1                       NaN                                 0.879220   

   android.sensor.linear_acceleration#min  \
0                                0.002495   
1                                0.641117   

   android.sensor.linear_acceleration#max  \
0                                 0.05241   
1                                 1.18581   

   android.sensor.linear_acceleration#std  android.sensor.magnetic_field#mean  \
0                                0.011045                           57.099638   
1                                0.278730                           29.351288   

   android.sensor.magnetic_field#min  android.sensor.magnetic_field#max  \
0                          56.690387                          57.575950   
1                          28.172505                          30.386017   

   android.sensor.magnetic_field#std  \
0                           0.177549   
1                           0.921547   

   android.sensor.magnetic_field_uncalibrated#mean  

In [3]:
# カラム名をわかりやすさのため変更します
df.columns = df.columns.str.replace('android.sensor.','').str.replace('#','_')

/var/folders/x2/7ss5kcb12n1gtlpc8p2mtjtw0000gn/T/ipykernel_65166/2509734072.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  df.columns = df.columns.str.replace('android.sensor.','').str.replace('#','_')


ここでは、移動手段（Car, Still, Train, Bus, Walking）を次の変数を用いて推定するモデルを考えます。

* `accelerometer_mean`
* `game_rotation_vector_mean`
* `gyroscope_mean`
* `linear_acceleration_mean`
* `orientation_mean` 
* `pressure_mean`
* `rotation_vector_mean`
* `sound_mean`

上の変数の欠損している行を分析から除外します。

In [4]:
df = df[[
    'target','accelerometer_mean','game_rotation_vector_mean',
    'gyroscope_mean','linear_acceleration_mean','orientation_mean', 
    'pressure_mean','rotation_vector_mean','sound_mean'
]]
df = df.dropna()

In [5]:
df.describe()

accelerometer_mean  game_rotation_vector_mean  gyroscope_mean  \
count         1255.000000                1255.000000     1255.000000   
mean             9.996223                   0.732083        0.325744   
std              0.720097                   0.246679        0.496171   
min              7.369055                   0.042684        0.001164   
25%              9.738144                   0.570153        0.027239   
50%              9.830731                   0.763646        0.083699   
75%             10.049428                   0.969717        0.512946   
max             14.661139                   0.999999        3.268860   

       linear_acceleration_mean  orientation_mean  pressure_mean  \
count               1255.000000       1255.000000    1255.000000   
mean                   1.507555        208.533199    1020.577142   
std                    1.890074         92.273671      15.173493   
min                    0.009447         17.135360     946.339840   
25%                    0.272564        135.326907    1011.055625   
50%                    0.707634        206.208459    1020.669000   
75%                    2.083757        291.281798    1029.123300   
max                   18.141770        396.133498    1058.609900   

       rotation_vector_mean   sound_mean  
count           1255.000000  1255.000000  
mean               0.768484    75.202876  
std                0.224416    12.239499  
min                0.120813     0.000000  
25%                0.668633    71.869349  
50%                0.813465    77.707224  
75%                0.959668    82.794433  
max                1.000000    90.308734

In [7]:
x = df.drop('target', axis=1)
y = df['target']

In [8]:
x.corr()

accelerometer_mean  game_rotation_vector_mean  \
accelerometer_mean                   1.000000                   0.129695   
game_rotation_vector_mean            0.129695                   1.000000   
gyroscope_mean                       0.500263                   0.297029   
linear_acceleration_mean             0.441453                   0.395925   
orientation_mean                     0.043532                   0.215494   
pressure_mean                       -0.045848                  -0.144826   
rotation_vector_mean                 0.081403                   0.488242   
sound_mean                           0.138451                   0.179971   

                           gyroscope_mean  linear_acceleration_mean  \
accelerometer_mean               0.500263                  0.441453   
game_rotation_vector_mean        0.297029                  0.395925   
gyroscope_mean                   1.000000                  0.732075   
linear_acceleration_mean         0.732075                  1.000000   
orientation_mean                 0.050534                  0.096761   
pressure_mean                   -0.032837                 -0.099997   
rotation_vector_mean             0.169035                  0.226298   
sound_mean                       0.017961                  0.072577   

                           orientation_mean  pressure_mean  \
accelerometer_mean                 0.043532      -0.045848   
game_rotation_vector_mean          0.215494      -0.144826   
gyroscope_mean                     0.050534      -0.032837   
linear_acceleration_mean           0.096761      -0.099997   
orientation_mean                   1.000000       0.007241   
pressure_mean                      0.007241       1.000000   
rotation_vector_mean              -0.007124      -0.288926   
sound_mean                         0.016743       0.054731   

                           rotation_vector_mean  sound_mean  
accelerometer_mean                     0.081403    0.138451  
game_rotation_vector_mean              0.488242    0.179971  
gyroscope_mean                         0.169035    0.017961  
linear_acceleration_mean               0.226298    0.072577  
orientation_mean                      -0.007124    0.016743  
pressure_mean                         -0.288926    0.054731  
rotation_vector_mean                   1.000000   -0.010011  
sound_mean                            -0.010011    1.000000

In [9]:
y.value_counts().to_frame()

target
Walking     346
Train       311
Car         310
Bus         220
Still        68

In [10]:
x = sm.add_constant(x, prepend = False)
mnlogit_mod = sm.MNLogit(y, x,
                         # method="bfgs", 
                         # maxiter=250
                        )
mnlogit_fit = mnlogit_mod.fit()
print (mnlogit_fit.summary())

Optimization terminated successfully.
         Current function value: 0.488472
         Iterations 11
                          MNLogit Regression Results                          
Dep. Variable:                 target   No. Observations:                 1255
Model:                        MNLogit   Df Residuals:                     1219
Method:                           MLE   Df Model:                           32
Date:                Wed, 17 May 2023   Pseudo R-squ.:                  0.6764
Time:                        11:16:00   Log-Likelihood:                -613.03
converged:                       True   LL-Null:                       -1894.5
Covariance Type:            nonrobust   LLR p-value:                     0.000
               target=Car       coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
accelerometer_mean           -1.3405      0.326     -4.117      0.000      -1

多項ロジットモデルでは、選択肢の数をkとするとk-1個のモデルを推定します。
上の出力を見ると、5つの選択肢のうちBusがベース（参照グループ）として設定されているため、出力された結果を見る際には、パラメーターの推定値は、参照グループであるバスに対する相対的なものとなります。

多項ロジットモデルでの一般的な解釈は、説明変数の単位変化に対して、参照グループに対して、モデル内の他の変数が一定である場合、パラメーター推定値の分だけ変化すると期待されることになります。

上の結果で`target=Car`の出力結果では、バスと車の相対比較を示しています。

`accelerometer_mean`の推定された係数は`-1.340`で、モデル内の他の変数が一定である場合、バスに対する車の`accelerometer_mean`の値が1単位増えると、バスより車である可能性は1.340減少すると解釈できます。

同様に、`game_rotation_vector_mean`が1単位増えると、モデル内の他の変数が一定である場合、ユーザーの移動手段はバスであるよりも車である可能性が2.8793減少する可能性があると解釈できます。